### PR TITLE
reducing the amount of configs needed.

### PR DESCRIPTION
**Pull request by borissmidt**
I think it still needs some work but it at least reduces the amount of these closures. Since it slows down the dependency resolution a lot. i..e for all calls of all dependnecies every closure is executed.
The amount of these registrations should be reduced and i think those big sets of dependencies is what eventually makes gradle crawl.
I.e. i got a OOM when i had a project with 30 sub projects.

This already reduces it by a square root of amount. i.e. before it was applied for every project by every project and for all configs.
so for 30 projects with more then 20 configurations each it becomes huge very fast. 

It might be that the deamon doesn't clean the state and thus creates a memory leak.

could you have another look on how we can reduce the amount of dependency rules.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -119,15 +119,15 @@ uploadArchives.repositories.mavenDeployer {
...
 }
 
-codenarc {
-    toolVersion = "1.1"
-}
-
-codenarcMain {
-    configFile = rootProject.file("config/codenarc/CodenarcMain.groovy")
-}
+//codenarc {
```
Comment (borissmidt) [inline build.gradle]: I need to renable this and remvoe some code, 
But i first want to get a good fix before that.

Comment (borissmidt) [reply build.gradle]: for me it fails the build that's why i disabled it.

File: src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy [hunk 2]
```diff
@@ -27,13 +27,16 @@ import org.gradle.api.artifacts.Configuration
 import org.gradle.api.artifacts.DependencyResolveDetails
 import org.gradle.api.artifacts.DependencySet
 
+import java.util.concurrent.atomic.AtomicLong
```
Comment (prokod) [inline src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy]: @borissmidt do we need this still ?

File: src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy [hunk 3]
```diff
@@ -59,7 +62,7 @@ class ResolutionStrategyConfigurer {
      *                       - A {@link Configuration} to link as {@code extendedFrom}
      */
     void applyForLinkWith(ViewType... views) {
-        def project = sourceSetInsights.project
+
```
Comment (prokod) [inline src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy:64]: There is another occurrence that needs removal

File: src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy [hunk 4]
```diff
@@ -69,46 +72,49 @@ class ResolutionStrategyConfigurer {
...
-                    lifecycle:'afterEvaluate',
-                    configuration:crossBuildConfigurationName,
-                    parentConfiguration:names.main,
-                    msg:"Inherited dependendencies to consider while resolving ${crossBuildConfigurationName} " +
-                            'configuration dependencies: [' +
-                            "${allDependencies.collect { "${it.group}:${it.name}" }.join(', ')}]"
-            ))
 
+            //avoid big string creation
+            if (project.logger.infoEnabled) {
```
Comment (prokod) [inline src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy:76]: What is the performance hit when info is enabled?

File: src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy [hunk 5]
```diff
@@ -69,46 +72,49 @@ class ResolutionStrategyConfigurer {
...
+                    configuration: crossBuildConfigurationName,
+                    parentConfiguration: names.main,
+                    msg: "Inherited dependendencies to consider while resolving ${crossBuildConfigurationName} " +
+                        'configuration dependencies: [' +
+                        "${allDependencies.collect { "${it.group}:${it.name}" }.join(', ')}]"
+                ))
+            }
             def di = new DependencyInsights(sourceSetInsights)
 
-            def crossBuildProjects = CrossBuildPluginUtils.findAllCrossBuildPluginAppliedProjects(insightsView)
```
Comment (prokod) [inline src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy:83]: Why did you remove that ?

**Review by prokod — COMMENTED** [review]
Thx @borissmidt for the PR. Please see my comments/questions

File: src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy (hunk 4 above)
Comment (borissmidt) [reply src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy:76]: I don't know exact but making a string with all dependencies probably gives some extra GC overhead.

File: src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy (hunk 5 above)
Comment (borissmidt) [reply src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy:83]: this was the main problem that for each project you run this funciton and then search all other projects with cross build enabled and then configure the rules for each project with crossbuild enabled.

i.e. 30 projects find the other 30 projects which results in 900 times that these are configured while it only needs to be configured 30 times.

File: src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy [hunk 6]
```diff
@@ -69,46 +72,49 @@ class ResolutionStrategyConfigurer {
...
             def unionOfAllDependenciesAsDisplayNameSet =
-                    unionOfAllDependencies.collect { dep -> "${dep.group}:${dep.name}:${dep.version}" }.toSet()
+                unionOfAllDependencies.collect { dep -> "${dep.group}:${dep.name}:${dep.version}" }.toSet()
 
-            crossBuildProjects.each {
-                it.configurations.all { Configuration c ->
-                    c.resolutionStrategy.eachDependency { details ->
-                        resolutionStrategyHandler(c, details, unionOfAllDependenciesAsDisplayNameSet, view)
+            project.configurations.each { Configuration c ->
+                    if (c.name.startsWith("crossBuild")){
```
Comment (borissmidt) [inline src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy:96]: i think this is an important change and i wonder if it is correct.

i.e. now it was also looking at a lot of other configuratio where it didn't need to reconfigure anything. i.e. zinc configuration, testCompile compile etc.

File: src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy (hunk 2 above)
Comment (borissmidt) [reply src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy]: fixed

File: src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy (hunk 3 above)
Comment (borissmidt) [reply src/main/groovy/com/github/prokod/gradle/crossbuild/ResolutionStrategyConfigurer.groovy:64]: fixed

**Review by borissmidt — COMMENTED** [review]
BTW i tested it on the project where i had the issue and found the 2.7M instances of the closure and it is significantly reduced.
So i don't have these OOM's anymore.

**Comment by borissmidt** [comment]
Could you please accept this PR because we have some builds failing because they go out of memory.
In the meantime we will do a forked release internally.

**Comment by prokod** [comment]
Hi @borissmidt I have some busy times this days. Sorry for not contacting earlier. Please bare with me. basically I was able to set up a test scenario for this which I wouldlike to integrate first before I make a release. As I said this work got halted. I will return to this hopefully by end of week.

Thx and really appreciated!

**Comment by prokod** [comment]
@borissmidt I have some more room for improvements after further debug. I will create a PR for that based on your branch.

**Comment by borissmidt** [comment]
Cool that you are picking it up, the plugin is getting better and better :)

**Comment by prokod** [comment]
This PR is superseded by #105
